### PR TITLE
[ENH] `torch`-based CNTC time-series classifier and regressor

### LINKED ISSUE #9319: [ENH] `torch` based cntc time series classifier and regressor
## Is your feature request related to a problem? Please describe.
This is a sub-issue of #8699 
Implement cntc time series classifier and regressor in `torch`, based on the old `tensorflow` implementations which can be found in `sktime/sktime/networks/`, `sktime/sktime/classification/deep_learning/` and `sktime/sktime/regressor/deep_learning/`

## Describe the solution you'd like
[SimpleRNNClassifierTorch](https://github.com/sktime/sktime/blob/main/sktime/classification/deep_learning/rnn/_rnn_torch.py) implemented in [PR/#8842](https://github.com/sktime/sktime/pull/8842) can be used as a template for classifier migrations. Similarly, [SimpleRNNRegressorTorch](https://github.com/sktime/sktime/blob/main/sktime/regression/deep_learning/rnn/_rnn_torch.py) implemented in [PR/#9013](https://github.com/sktime/sktime/pull/9013) can be used as a template for regressor migrations.
Discussion in comment section of both the above mentioned PRs gives an idea behind certain choices made and also possible improvements that can be made, should you need to refer.

Please include the following checklist in your PR.
- [ ] Implemented the network structure in `sktime/sktime/networks/`
- [ ] Implemented the classifier in `sktime/sktime/classification/deep_learning/`
- [ ] Implementded the regressor in `sktime/sktime/regressor/deep_learning/`
- [ ] Followed the naming convention and directory structure from RNN migration samples
- [ ] Followed the [sktime docstring convention](https://www.sktime.net/en/latest/developer_guide/documentation.html#docstring-conventions). Please refer the RNN migration samples to understand the usage.
- [ ] Added meaningful test parameters in the classmethod `get_test_params` for classifiers and regressors.
- [ ] Followed the [sktime guide on implementing estimators](https://www.sktime.net/en/latest/developer_guide/add_estimators.html)
- [ ] Used same defaults in `torch` implemention as used in the corresponding `tensorflow` implementation of same model.

## Additional context
[Issue/#8928](https://github.com/sktime/sktime/issues/8928) consolidates further improvements/enhancements that are desirable and will be appreciated if you choose to incorporate. Feel free to brainstorm or share your ideas on these.

### PR DESCRIPTION
**Pull request by srupat**
#### Reference Issues/PRs
Closes #9319. Completes PR #9397 by @fnhirwa. 

#### What does this implement/fix? Explain your changes.
Implements the CNTC classifier and regressor in PyTorch. Uses a separate network class to build the network. 
Follows the best practices and design decisions discussed and implemented in #8842 and #9013. Also review comments from #9412.

#### Does your contribution introduce a new dependency? If yes, which one?
No.

#### What should a reviewer concentrate their feedback on?
* 3d981b9 : Review the network class for logic. The logic follows the paper closely. Additional flexibility in terms of addition of layers at each stage is given. Also check `get_test_params` in the classifier and regressor for appropriate test coverage.
* 91f73f3 : `SeqSelfAttentionTorch` had intensity and attention as non-leaf tensors which torch flags when copying since it can only copy user created attributes. Added `.detach()` for these (they are anyways kept for observability/debugging purposes). This also helps pass the deepcopy related tests in CNTC and TapNet torch estimators. 
* 7e492cf : regressor base torch class has a typo: should use `optimizer_kwargs` instead of `criterion_kwargs`. (Not sure whether this belongs in a seperate PR, @fkiraly LMK if I should open a seperate PR for this)

#### Did you add any tests for the change?
No. All local `check_estimator` tests pass. 

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/allcontributors.org/blob/main/src/content/docs/en/reference/emoji-key.mdx)
- [X] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [X] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [X] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [X] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency)

### CODE REVIEW

**Review by fkiraly — APPROVED** [review]
Looks good. The `CNTClassifier` failures are unrelated (to the keras version)
